HDDS-6367. ContainerBalancer shows incorrect iteration result sometimes

siddhantsangwan opened the pull request:
## What changes were proposed in this pull request?

Balancer show an incorrect iteration result in some cases, such as:
1. Reaches the limit for max size to move in an iteration, but some moves fail (in replication manager). This will return the incorrect result `MAX_SIZE_TO_MOVE_REACHED`. The correct result should be `ITERATION_COMPLETED`.
2. Same as above, but reaches the limit for max datanodes to involve and then some moves fail in replication manager.

This PR changes the `doIteration` and `checkIterationMoveResults` methods in ContainerBalancer to solve this bug.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6367

## How was this patch tested?

Added a UT and some assertions in the existing ones.

siddhantsangwan commented:
@lokeshj1703 @JacksonYao287 can you please review?

symious commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:
```diff
@@ -417,24 +425,32 @@ private IterationResult doIteration() {
-      }
-      return IterationResult.ITERATION_COMPLETED;
     } finally {
-      checkIterationMoveResults(selectedTargets);
```
I think finally block better fits the idea of cleaning code, the new version seems to have too much process logic in it. Personally, I do like the original method.

siddhantsangwan replied:
Yeah, but the original method logic is buggy. I think we can remove the finally block entirely if we don't return early from the `doIteration` method. If we have instead just break out from the while loop and return only at the end after processing results, we don't need try-finally.

symious replied:
I agree, the ITERATION_RESULT is a little confused here. It's now referring to the 
1. reason for stop generating moves (MAX_SIZE_TO_MOVE_REACHED)
2. the result of inner iteration (ITERATION_COMPLETED)
3. the reason we should stop the whole balancer service (CAN_NOT_BALANCE_ANY_MORE)

Maybe we should sort out the process and refactor the code.

symious commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:
```diff
@@ -375,26 +378,31 @@ private IterationResult doIteration() {
-        IterationResult result = checkConditionsForBalancing();
-        if (result != null) {
-          return result;
+        if (checkConditionsForBalancing() != null) {
```
I think the `maxSizeToMovePerIteration` and `maxDatanodesPercentageToInvolvePerIteration` are more like ITERATION_STOP_REASON other than ITERATION_RESULT.
Maybe we can add a new field noting the reason and mark the result as COMPLETE if encounter a valid STOP_REASON?

siddhantsangwan commented:
@symious Thanks for the review!

> I think the maxSizeToMovePerIteration and maxDatanodesPercentageToInvolvePerIteration are more like ITERATION_STOP_REASON other than ITERATION_RESULT.
Maybe we can add a new field noting the reason and mark the result as COMPLETE if encounter a valid STOP_REASON?

Makes sense. This is an alternative that I was thinking of - We can just mark the result as `ITERATION_COMPLETED` when the max size or max DNs limit is hit. 

Should we have only two possible results for an iteration, then? Such as `ITERATION_COMPLETED` and `ITERATION_STOPPED`. The rest can be logged as reasons for stopping the iteration as you're suggesting.

symious commented:
> Should we have only two possible results for an iteration, then? Such as ITERATION_COMPLETED and ITERATION_STOPPED. The rest can be logged as reasons for stopping the iteration as you're suggesting.

If we generated 10 moves, of which 5 succeeded, 5 failed. I was wondering which result should this case fit?

siddhantsangwan commented:
> > Should we have only two possible results for an iteration, then? Such as ITERATION_COMPLETED and ITERATION_STOPPED. The rest can be logged as reasons for stopping the iteration as you're suggesting.
> 
> If we generated 10 moves, of which 5 succeeded, 5 failed. I was wondering which result should this case fit?

It should be `ITERATION_COMPLETED` according to me. The move failures can be logged.

symious commented:
> It should be ITERATION_COMPLETED according to me. The move failures can be logged.

The total and failed moves can be added to metrics for better monitoring.

siddhantsangwan commented:
I've simplified the result to three scenarios:

1. `ITERATION_INTERRUPTED` if balancing was stopped while iterating.
2. `CAN_NOT_BALANCE_ANY_MORE` if no move was generated during this iteration.
3. `ITERATION_COMPLETED` for all other cases, such as when some (or all) container moves failed or limits were hit.

@symious I've also removed the try...finally block since we won't return early from the `doIteration` method now.

symious commented:
LGTM.
Thank you for the update. @siddhantsangwan

siddhantsangwan commented:
@symious Thanks for reviewing. I've created a JIRA for adding failed moves metrics: https://issues.apache.org/jira/browse/HDDS-6492

lokeshj1703 commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:
```diff
@@ -408,57 +410,77 @@ private IterationResult doIteration() {
+       If no move was generated during this iteration then we don't need to
+       check the move results
+       */
+      iterationResult = IterationResult.CAN_NOT_BALANCE_ANY_MORE;
```
I think there is a case which we might need to handle. Suppose for a selected source datanode, no target datanode/container pair can be found. In such  case balancer will terminate as moveGenerated is false but ideally we should exhaust all source datanodes.

lokeshj1703 replied:
I realized this would not be the case as iteration results are checked after the while loop.

lokeshj1703 reviewed (commented):
@siddhantsangwan Thanks for working on this! The changes look good. I have a comment inline.

lokeshj1703 commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:
```diff
@@ -408,57 +410,77 @@ private IterationResult doIteration() {
-        //balanced any more, just stop iteration and exit
-        return IterationResult.CAN_NOT_BALANCE_ANY_MORE;
+      ContainerMoveSelection moveSelection = matchSourceWithTarget(source);
+      if (moveSelection != null) {
+        isMoveGenerated = true;
+        LOG.info("ContainerBalancer is trying to move container {} from " +
+                "source datanode {} to target datanode {}",
+            moveSelection.getContainerID().toString(),
+            source.getUuidString(),
+            moveSelection.getTargetNode().getUuidString());
+
+        if (moveContainer(source, moveSelection)) {
+          // consider move successful for now, and update selection criteria
+          updateTargetsAndSelectionCriteria(
+              selectedTargets, moveSelection, source);
+        }
+      } else {
+        // can not find any target for this source
+        findSourceStrategy.removeCandidateSourceDataNode(source);
       }
```
Can we move the moveSelection processing logic to a separate function?

lokeshj1703 commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java:
```diff
@@ -408,57 +410,77 @@ private IterationResult doIteration() {
       }
-      return IterationResult.ITERATION_COMPLETED;
-    } finally {
+    }
```
Better to add a new line after while loop.

siddhantsangwan commented:
@lokeshj1703 thanks for the review. Updated.

lokeshj1703 reviewed (approved):
@siddhantsangwan Thanks for updating the PR! Pending clean CI.

siddhantsangwan commented:
@lokeshj1703 Checks passed now!

lokeshj1703 commented:
@siddhantsangwan Thanks for the contribution! @symious Thanks for the review! I have committed the PR to master branch.
